### PR TITLE
Improve the proxy authenticator

### PR DESCRIPTION
**Pull request by twasyl**
Signed-off-by: Thierry Wasylczenko <thierry.wasylczenko@gmail.com>

Until now, the `JenkinsProxyAuthenticator` checked the authentication scheme in a case sensitive way. This check has been updated to ignore the case.

Moreover, it seems OkHttp always tries to send an authentication request using the `Proxy-Authenticate: OkHttp-Preemptive` prior to any `CONNECT` request. According the [documentation](https://square.github.io/okhttp/4.x/okhttp/okhttp3/-authenticator/)

> Prior to sending any CONNECT request OkHttp always calls the proxy authenticator so that it may prepare preemptive authentication. OkHttp will call [authenticate](https://square.github.io/okhttp/4.x/okhttp/okhttp3/-authenticator/authenticate/) with a fake HTTP/1.1 407 Proxy Authentication Required response that has a Proxy-Authenticate: OkHttp-Preemptive challenge. The proxy authenticator may return either an authenticated request, or null to connect without authentication.

This is why the authentication scheme also checks for this header as well. 

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/okhttp/api/internals/JenkinsProxyAuthenticator.java [hunk 1]
```diff
@@ -47,4 +48,18 @@ public Request authenticate(@Nullable Route route, Response response) throws IOE
...
+
+        /* According https://square.github.io/okhttp/4.x/okhttp/okhttp3/-authenticator/ :
+
+             | OkHttp will call authenticate with a fake HTTP/1.1 407 Proxy Authentication Required response that has a
+             | Proxy-Authenticate: OkHttp-Preemptive challenge. The proxy authenticator may return either an
+             | authenticated request, or null to connect without authentication.
+
+           That is why we check for the OkHttp-Preemptive header.
+         */
+        return lowerCaseHeader.startsWith("basic") || lowerCaseHeader.equalsIgnoreCase("okhttp-preemptive");
```
Comment (jtnord) [inline src/main/java/io/jenkins/plugins/okhttp/api/internals/JenkinsProxyAuthenticator.java:63]: you have already lower cased this - why use `equalsIgnoreCase` ?

File: src/main/java/io/jenkins/plugins/okhttp/api/internals/JenkinsProxyAuthenticator.java [hunk 2]
```diff
@@ -47,4 +48,18 @@ public Request authenticate(@Nullable Route route, Response response) throws IOE
 
         return null;
     }
+
+    private boolean isAuthenticationSchemeSupported(@NonNull final String proxyAuthenticateHeader) {
+        final String lowerCaseHeader = proxyAuthenticateHeader.toLowerCase();
```
Comment (jtnord) [inline src/main/java/io/jenkins/plugins/okhttp/api/internals/JenkinsProxyAuthenticator.java:53]: lower case of `BASIC` is not always `basic` and so this code will not always work.

please ensure the use of `Locale.ROOT` and not rely on the locale that Jenkins is running in.

**Review by jtnord — CHANGES_REQUESTED** [review]

File: src/test/java/jenkins/plugins/okhttp/api/ClientTest.java [hunk 3]
```diff
@@ -164,6 +165,42 @@ public void testProxy() throws ExecutionException, InterruptedException, IOExcep
...
     }
 
+    @Test
+    public void testProxyWithBasicAuthAndRealm() throws ExecutionException, InterruptedException, IOException {
+        jenkinsRule.jenkins.setProxy(new ProxyConfiguration("127.0.0.1", proxy.port(), "proxy-user", "proxy-pass"));
+        secureProxy("Basic realm=\"somerealm\"", "proxy-user:proxy-pass");
+        server.stubFor(WireMock.get("/hello").willReturn(aResponse().proxiedFrom(proxy.baseUrl())));
+
+        final OkHttpClient client = JenkinsOkHttpClient.newClientBuilder(new OkHttpClient())
+                .build();
```
Comment (aHenryJard) [inline src/test/java/jenkins/plugins/okhttp/api/ClientTest.java:175]: this `.build();` seems a bit lonely, can you attach it to previous line or maybe do the break line with `.newClientBuilder` and aligned them vertically ?

File: src/test/java/jenkins/plugins/okhttp/api/ClientTest.java [hunk 4]
```diff
@@ -164,6 +165,42 @@ public void testProxy() throws ExecutionException, InterruptedException, IOExcep
...
+    }
+
+    @Test
+    public void testProxyWithPreemptiveAuth() throws ExecutionException, InterruptedException, IOException {
+        jenkinsRule.jenkins.setProxy(new ProxyConfiguration("127.0.0.1", proxy.port(), "proxy-user", "proxy-pass"));
+        secureProxy("OkHttp-Preemptive", "proxy-user:proxy-pass");
+        server.stubFor(WireMock.get("/hello").willReturn(aResponse().proxiedFrom(proxy.baseUrl())));
+
+        final OkHttpClient client = JenkinsOkHttpClient.newClientBuilder(new OkHttpClient())
+                .build();
```
Comment (aHenryJard) [inline src/test/java/jenkins/plugins/okhttp/api/ClientTest.java:193]: same remark for build()

File: src/main/java/io/jenkins/plugins/okhttp/api/internals/JenkinsProxyAuthenticator.java (hunk 1 above)
Comment (aHenryJard) [inline src/main/java/io/jenkins/plugins/okhttp/api/internals/JenkinsProxyAuthenticator.java:63]: with the toLowerCase() above, is the equalsIgnoreCase really needed here ? Is not equals enough ?

**Review by aHenryJard — APPROVED** [review]
With documentation and tests ! Awesome :D :heart:

File: src/main/java/io/jenkins/plugins/okhttp/api/internals/JenkinsProxyAuthenticator.java (hunk 1 above)
Comment (twasyl) [reply src/main/java/io/jenkins/plugins/okhttp/api/internals/JenkinsProxyAuthenticator.java]: Fixed

File: src/main/java/io/jenkins/plugins/okhttp/api/internals/JenkinsProxyAuthenticator.java (hunk 2 above)
Comment (twasyl) [reply src/main/java/io/jenkins/plugins/okhttp/api/internals/JenkinsProxyAuthenticator.java]: Fixed

File: src/test/java/jenkins/plugins/okhttp/api/ClientTest.java (hunk 3 above)
Comment (twasyl) [reply src/test/java/jenkins/plugins/okhttp/api/ClientTest.java]: Done

File: src/test/java/jenkins/plugins/okhttp/api/ClientTest.java (hunk 4 above)
Comment (twasyl) [reply src/test/java/jenkins/plugins/okhttp/api/ClientTest.java]: Done

File: src/main/java/io/jenkins/plugins/okhttp/api/internals/JenkinsProxyAuthenticator.java (hunk 1 above)
Comment (twasyl) [reply src/main/java/io/jenkins/plugins/okhttp/api/internals/JenkinsProxyAuthenticator.java]: Fixed

**Review by jtnord — APPROVED** [review]

File: src/test/java/jenkins/plugins/okhttp/api/ClientTest.java [hunk 5]
```diff
@@ -164,6 +165,40 @@ public void testProxy() throws ExecutionException, InterruptedException, IOExcep
...
+        assertEquals(200, response.code());
+        try (final ResponseBody body = response.body()) {
+            assertEquals("Hello from proxy", body.string());
+        }
+    }
+
+    @Test
+    public void testProxyWithPreemptiveAuth() throws ExecutionException, InterruptedException, IOException {
+        jenkinsRule.jenkins.setProxy(new ProxyConfiguration("127.0.0.1", proxy.port(), "proxy-user", "proxy-pass"));
+        secureProxy("OkHttp-Preemptive", "proxy-user:proxy-pass");
```
Comment (jtnord) [inline src/test/java/jenkins/plugins/okhttp/api/ClientTest.java:188]: this should not be seen by the proxy server so if this auth scheme is seen does it not imply something is broken?  (in other words this scheme should be internal to the okhttp API only IIUC)

File: src/test/java/jenkins/plugins/okhttp/api/ClientTest.java [hunk 6]
```diff
@@ -63,12 +62,14 @@ private void secureProxy(String authenticationType, final String... authorizedUs
...
                         .withHeader("Proxy-Authenticate", authenticationType)));
 
-        if (authorizedUserPass != null && "Basic".equals(authenticationType)) {
-            for (String userPass : authorizedUserPass) {
-                final String authenticationHeaderValue = "Basic " + Base64.getEncoder().encodeToString(userPass.getBytes(StandardCharsets.UTF_8));
-                proxy.stubFor(get(WireMock.anyUrl())
-                        .withHeader("Proxy-Authorization", WireMock.equalTo(authenticationHeaderValue))
-                        .willReturn(WireMock.ok("Hello from proxy")));
+        if (authorizedUserPass != null && authenticationType != null) {
+            if (authenticationType.toLowerCase().contains("basic") || authenticationType.equalsIgnoreCase("okhttp-preemptive")) {
```
Comment (jtnord) [inline src/test/java/jenkins/plugins/okhttp/api/ClientTest.java:66]:
```suggestion
            if (authenticationType.toLowerCase(Locale.ROOT).startsWith("basic")) {
```
see also discussion on the test.  this header should not be seen on the wire IIUC.

Comment (twasyl) [reply src/test/java/jenkins/plugins/okhttp/api/ClientTest.java:66]: Done.

File: src/test/java/jenkins/plugins/okhttp/api/ClientTest.java (hunk 5 above)
Comment (twasyl) [reply src/test/java/jenkins/plugins/okhttp/api/ClientTest.java]: Removed

**Review by jtnord — APPROVED** [review]
sure, but I was not trying to say do not support the *internal* pre-emtpive auth from OKHttp but that testing it by using a server that issues that challange is mostly an invalid IT as a server should never issue that scheme.

**Comment by twasyl** [comment]
> sure, but I was not trying to say do not support the internal pre-emtpive auth from OKHttp but that testing it by using a server that issues that challange is mostly an invalid IT as a server should never issue that scheme.

Sorry but you're suggestion removing the check for the authentication header `OkHttp-Preemptive` in `JenkinsProxyAuthenticator` lead me to think I should remove it.

Regarding the test, I do agree that a _real_ proxy will not issue this header, but I wanted to test that it's working well when this header is sent and it's covered by a test. So it made it clear that this was done on purpose, reason why I also put the OkHttp documentation in a comment. My goal here was not to make an integration test, but a unit one.

**Pull request merged**